2148 global snapshot calculation doesnt behave as expected

devyetii opened the pull request:
## Fixes
- #2148
## Changes
- Refactoring global snapshot calculation
- Adding unit tests for provider aggregate calculation + global snapshot updates

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test: 0chain/system_test#420
- zboxcli:
- zwalletcli:
- 0box: 0chain/0box#485

devyetii committed:
add provider count to global snapshot

devyetii committed:
use snapshot diff calculation in blobber aggregate

devyetii committed:
add provider count fields to snapshots table

devyetii committed:
use snapshot diff for providers snapshots

devyetii committed:
remove globalSnapshot

devyetii committed:
add snapshots test

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
complete snapshot test

devyetii committed:
add db partitions

devyetii committed:
fix snapshot unit tests

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix snapshot unit test

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/snapshot.go:
```diff
@@ -44,6 +43,63 @@ type Snapshot struct {
+
+	providerCount := s.providerCount(provider)
+	if providerCount > 0 {
+		s.AverageWritePrice += diff.AverageWritePrice / providerCount
```
it can be not correct if it is new blobber, 
previous average was calculated as `sum(wp)/n` and new one as `sum(wp + dif)/ (n+1)` so your share should be `(sum(wp)/n + diff/n)*n/(n+1)` and simply `old_sum * n / (n + 1) + diff / (n + 1)`

devyetii replied:
Nice catch 👌 Fixing it.

devyetii replied:
Fixed. FYI when new providers are added they affect the snapshot count directly as response to the events. https://github.com/0chain/0chain/blob/a89a68aaf0a4cba103f731b35997923d0c0f61a6/code/go/0chain.net/smartcontract/dbs/event/snapshot.go#L318-L354
I added update to the avg write price there to update it whenever a provider is added/removed through the formula `new_avg = total / new_count = old_avg * old_count / new_count`

devyetii replied:
Added some tests and just noticed that I don't need to update average for all the providers, just for the blobbers. Lack of coffee 🙃

devyetii committed:
update snapshot avgs after provider counts change

devyetii committed:
update average only for blobbers count change + unit tests

devyetii committed:
fix unit tests

devyetii committed:
blobber aggregate test

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix blobber aggregate unit test

fix blobber aggregate test

add blobber owner before test

fix blobber aggregate unit testing

fix unit test

fix unit test

fix unit test

add log

add miner aggregate unit test

fix blobber aggregate unit testing

add log

add sharder aggregate testing

fix blobber test

fix miner aggregate test

add log

add log

fix unit tests

test authorizer aggregate

fix unit test

add validator aggregate test + enhancement to others

fix validator unit test

fix unit testing

add logs

more logs

fix blobber aggregate test

fix blobber aggregate test

add log

fix blobber aggregate test

fix blobber aggregate test

add logs

add log

add log

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unti test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fxi unit test

manasnisar commented on code/go/0chain.net/smartcontract/dbs/goose/migrations/008_add_snapshot_provider_count.sql:
```diff
@@ -0,0 +1,17 @@
+
+-- +goose Down
+-- +goose StatementBegin
+ALTER TABLE allocations DROP COLUMN IF EXISTS miner_count;
```
Some nice work you did in this PR :)
I guess you wanted to drop columns from `snapshots` here

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
authorizer aggregate test

devyetii committed:
fix migration

devyetii committed:
miner/sharder aggregate tests

devyetii committed:
validator aggregate tests

devyetii committed:
remove unnecessery logs

devyetii commented:
@peterlimg @dabasov Kindly review

peterlimg commented on code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go:
```diff
@@ -100,6 +101,8 @@ func (edb *EventDb) calculateBlobberAggregate(gs *globalSnapshot, round, limit,
 	logging.Logger.Debug("blobber_snapshot", zap.Int("total_current_blobbers", len(currentBlobbers)))
 
 	if round <= edb.AggregatePeriod() && len(currentBlobbers) > 0 {
+		logging.Logger.Debug("creating snapshots", zap.Any("currentBlobbers", currentBlobbers))
```
@devyetii please check logs to see if this would cause log flush, the currentBlobbers could be a large log block.

devyetii replied:
was for testing, removed.

peterlimg commented on code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go:
```diff
@@ -141,28 +147,27 @@ func (edb *EventDb) calculateBlobberAggregate(gs *globalSnapshot, round, limit,
 		const GB = currency.Coin(1024 * 1024 * 1024)
 		if aggregate.WritePrice == 0 {
-			gs.StakedStorage = gs.MaxCapacityStorage
+			gs.StakedStorage = gs.MaxCapacityStorage	// chosed not to use gsDiff here not to over-complicate the calculation
```
Seems `gs.ApplyDiff(&gsDiff, spenum.Blobber)` can make the `gs.StakedStorage` overflow the `gs.MaxCapacityStorage`.

Suppose we have one aggregate.WritePrice is 0, that will set the `gs.StakedStorage` to max capacity. Then in another aggregate, its write price is normal, so we will have `gsDiff.StakedStorage += ss`. In `gs.ApplyDiff(&gsDiff...)` in line 169, if we don't do the max capacity checking, the `gs.StakedStorage` will get overflow.

devyetii replied:
Nice catch. Thanks Peter. Fixed.

devyetii committed:
resolve comments

devyetii committed:
add test to StakedStorage will not exceed MaxCapacityStorage

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
save totals instead of avgs

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix total txn update

devyetii committed:
all snapshot updates related to blobber should be in blobber_aggregate.go

devyetii committed:
more coverage of blobber snapshot cases

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix unit tests

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit tests

fix unit test

fix unit test

fix unit test

fix unit tests

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit tests

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

fix unit test

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
add blobbers_stake and remove total_write_price for global snapshot

devyetii committed:
fix benchmark test

devyetii commented:
@peterlimg @dabasov much changes since latest review. Kindly review again.

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix division by zero case

devyetii committed:
Merge branch '2148-global-snapshot-calculation-doesnt-behave-as-expected' of github.com:0chain/0chain into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix unit test

devyetii committed:
fix unit test

dabasov commented on code/go/0chain.net/smartcontract/dbs/event/authorizer_aggregate_test.go:
```diff
@@ -0,0 +1,324 @@
+
+		// Initial run. Should register snapshots and aggregates of authorizers in bucket
+		eventDb.updateAuthorizerAggregate(5, 10, &initialSnapshot)
+		eventDb.Store.Get().Exec("DROP TABLE IF EXISTS authorizer_temp_ids")
```
don't need it, table is unique for each transaction

devyetii replied:
I need it bcz I do two successive calls to the function that creates it in the same test. If I deleted it then the second call will no create the table and will use the one from the prev run, which has some providers that I deleted and doesn't contain ones that I added.

dabasov reviewed (approved):
lgtm

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix storage_stake unit test

devyetii committed:
Merge branch '2148-global-snapshot-calculation-doesnt-behave-as-expected' of github.com:0chain/0chain into 2148-global-snapshot-calculation-doesnt-behave-as-expected

dabasov commented:
merge after #2178

devyetii committed:
fix staked storage calculation with high write price

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
fix write price calculation

devyetii committed:
fix read data update

devyetii committed:
fix unstake total update

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected

devyetii committed:
remove total_value_locked

devyetii committed:
remove total_value_locked

devyetii committed:
remove blobbers_stake and move total_stake to provider aggregates

devyetii committed:
test total_staked change

devyetii committed:
Merge branch 'staging' into 2148-global-snapshot-calculation-doesnt-behave-as-expected
